src/recon_buildblock/PLSPrior.cxx

```diff
@@ -0,0 +1,663 @@
+//
+//
+/*
+    Copyright (C) 2018 University of Leeds and University College of London
+                  2000- 2011, Hammersmith Imanet Ltd
+    This file is part of STIR.
+
+    This file is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+
+    This file is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    See STIR/LICENSE.txt for details
+*/
+/*!
+  \file
+  \ingroup priors
+  \brief  implementation of the stir::PLSPrior class
+
+  \author Daniel Deidda
+  \author Tsai Yu Jung
+
+*/
+
+#include "stir/recon_buildblock/PLSPrior.h"
+#include "stir/Succeeded.h"
+#include "stir/DiscretisedDensityOnCartesianGrid.h"
+#include "stir/IndexRange3D.h"
+#include "stir/IO///write_to_file.h"
+#include "stir/IO/read_from_file.h"
+#include "stir/is_null_ptr.h"
+#include "stir/info.h"
+#include <algorithm>
+using std::min;
+using std::max;
+
+/* Pretty horrible code because we don't have an iterator of neigbhourhoods yet
+ */
+
+START_NAMESPACE_STIR
+
+template <typename elemT>
+void
+PLSPrior<elemT>::initialise_keymap()
+{
+  base_type::initialise_keymap();
+  this->parser.add_start_key("PLS Prior Parameters");
+  this->parser.add_key("only 2D", &only_2D);
+  this->parser.add_key("eta", &eta);
+  this->parser.add_key("alpha", &alpha);
+  this->parser.add_key("kappa filename", &kappa_filename);
+  this->parser.add_key("anatomical_filename", &anatomical_filename);
+  this->parser.add_key("weights", &weights);
+  this->parser.add_key("gradient filename prefix", &gradient_filename_prefix);
+  this->parser.add_stop_key("END PLS Prior Parameters");
+}
+
+
+template <typename elemT>
+bool
+PLSPrior<elemT>::post_processing()
+{
+  if (base_type::post_processing()==true)
+    return true;
+  if (kappa_filename.size() != 0)
+    this->kappa_ptr = read_from_file<DiscretisedDensity<3,elemT> >(kappa_filename);
+
+  if (anatomical_filename.size() != 0){
+    this->anatomical_sptr = read_from_file<DiscretisedDensity<3,elemT> >(anatomical_filename);}
+  else{
+      this->anatomical_sptr->fill (0);
+  }
+
+
+  if (this->anatomical_filename != "0"){
+      info(boost::format("Reading MR data '%1%'")
+           % anatomical_filename  );}
+
```

Comment:
please split this function into 2: from below a `Succeeded set_up()`. Then call this in `post_processing()`. This way, we'll be able to set parameters interactively first, call `set_up` and then go. (i.e. iwthout parsing).

Reply:
Ok, so whatever comes after line 83 should go in set_up? Also, I am removing everything related to "weights" as it is not used if that's ok.

Reply:
yes. The idea is that `post_processing` takes care of parsing related keywords. `set_up` does all the rest. `post_processing` will not be called if you don't parse anything (e.g. from Python)